### PR TITLE
Initial configuration that runs full closed loop with cwfs.

### PR DESCRIPTION
**Pull request by jbkalmbach**

### CODE REVIEW

File: policy/phosimCmptSetting.yaml [hunk 1]
```diff
@@ -13,6 +13,10 @@ intraDirName: intra
 # camera piston.
 extraDirName: extra
 
+# Wavefront Sensor directory name. This is needed if the defocal image is
+# provided by the wavefront sensors.
+wfsDirName: wfs
```
Comment (teweitsai) [inline policy/phosimCmptSetting.yaml:19]: If this is the directory name, where is the path (or root) in the file system? Or this is compared with the root of this repository?

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 2]
```diff
@@ -715,6 +745,58 @@ def _calcWfErrFromImg(
             Observation ID used in PhoSim.
         butlerRootPath : str
             Path to the butler repository.
+        instName : `str`
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: instName : `str` --> instName : str

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 3]
```diff
@@ -715,6 +745,58 @@ def _calcWfErrFromImg(
...
+    ):
+        """Calculate the wavefront error from the images generated by PhoSim.
+
+        Parameters
+        ----------
+        obsId : int
+            Observation ID used in PhoSim.
+        butlerRootPath : str
+            Path to the butler repository.
+        instName : `str`
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: instName : `str` --> instName : str

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 4]
```diff
@@ -753,14 +835,98 @@ def _calcWfErrFromImg(
...
+        butlerRootPath : `str`
+            Path to the butler gen3 repos.
+        instName : `str`
+            Instrument name.
+        numPro : int, optional
+            Number of processor to run DM pipeline. (the default is 1.)
+
+        Returns
+        -------
+        listOfWfErr : `list` of `SensorWavefrontError`
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py]: Please be consistent that you want to add `` to data type or not (see `numPro`).

File: python/lsst/ts/phosim/CloseLoopTask.py [hunk 5]
```diff
@@ -1013,7 +1194,15 @@ def runImg(
                 pathSkyFile=pathSkyFile,
             )
 
-        self.phosimCmpt.tele.setInstName(camType)
+        if inst == "lsst":
+            # Append equal weights for CWFS fields to OFC data
+            self.phosimCmpt.tele.setInstName(camType, defocalDist=0.0)
+            self.ofcCalc.ofc_data.normalized_image_quality_weight = np.append(
+                self.ofcCalc.ofc_data.normalized_image_quality_weight,
+                [0.25, 0.25, 0.25, 0.25],
```
Comment (teweitsai) [inline python/lsst/ts/phosim/CloseLoopTask.py:1210]: Please leave a comment to explain this magic number: [0.25, 0.25, 0.25, 0.25]. In the future, we may want to put this into the configuration file or use a new algorithm to adjust the weighting in the run time.

File: python/lsst/ts/phosim/OpdMetrology.py [hunk 6]
```diff
@@ -195,6 +195,16 @@ def setDefaultLsstGQ(self):
         )
         self.setWgtAndFieldXyOfGQ("lsst")
 
+    def setDefaultLsstWfsGQ(self):
+        """Set default values for LSST WFS field X, Y
+        and weighting ratio.
+        """
+
+        # Set Equal Weights for WFS
+        self.setWeightingRatio([1.0, 1.0, 1.0, 1.0])
```
Comment (teweitsai) [inline python/lsst/ts/phosim/OpdMetrology.py:205]: Please have a comment to explain: `[1.0, 1.0, 1.0, 1.0]`.

File: python/lsst/ts/phosim/PhosimCmpt.py [hunk 7]
```diff
@@ -715,6 +729,79 @@ def getPistonCamStarArgsAndFilesForPhoSim(
...
+            Star instance file name. (the default is "star.inst".)
+        logFileName : str, optional
+            Log file name. (the default is "starPhoSim.log".)
+        simSeed : int, optional
+            Random number seed. (the default is 1000)
+        cmdSettingFileName : str, optional
+            Physical command setting file name. (the default is
+            "starDefault.cmd".)
+        instSettingFileName : str, optional
+            Instance setting file name. (the default is "starSingleExp.inst".)
```
Comment (teweitsai) [inline python/lsst/ts/phosim/PhosimCmpt.py]: The order should be `obsId`, `skySim`, `simSeed`, `cmdSettingFileName`, and `instSettingFileName`. In addition, you should not have `cmdFileName`, `instFileName`, and `logFileName`.

File: python/lsst/ts/phosim/PhosimCmpt.py [hunk 8]
```diff
@@ -1198,6 +1288,45 @@ def getListOfFwhmSensorData(self, pssnFileName, sensorIdList):
 
         return fwhmCollection, sensor_id
 
+    def repackageWfsCamImgs(self, instName, isEimg=False):
+        """Repackage the images from in focus camera for processing.
+
+        Parameters
+        ----------
+        instName : `str`
```
Comment (teweitsai) [inline python/lsst/ts/phosim/PhosimCmpt.py]: instName : `str` --> instName : str

File: python/lsst/ts/phosim/PhosimCmpt.py [hunk 9]
```diff
@@ -1198,6 +1288,45 @@ def getListOfFwhmSensorData(self, pssnFileName, sensorIdList):
...
+        wfsDirName = self.getWfsDirName()
+
+        # Repackage the images to the temporary directory
+        command = "phosim_repackager.py"
+        phosimImgDir = os.path.join(self.outputImgDir, wfsDirName)
+        argstring = "%s --out_dir=%s" % (phosimImgDir, tmpDirPath)
+        argstring += f" --inst {instName} "
+        if isEimg:
+            argstring += " --eimage"
+        argstring += " --focusz 0"
```
Comment (teweitsai) [inline python/lsst/ts/phosim/PhosimCmpt.py:1310]: You may want to explain the value `0` here.

File: python/lsst/ts/phosim/telescope/TeleFacade.py [hunk 10]
```diff
@@ -324,14 +324,22 @@ def setInstName(self, camType, defocalDist=None):
...
+        ValueError
+            Defocal distance can not be < 0.
         """
 
         if defocalDist is None:
             defocalDist = self.getDefaultDefocalDist()
 
-        if defocalDist <= 0:
-            raise ValueError("Defocal distance can not be <= 0.")
+        if (defocalDist == 0) & (camType != CamType.LsstCam):
```
Comment (teweitsai) [inline python/lsst/ts/phosim/telescope/TeleFacade.py]: `if (defocalDist == 0) & (camType != CamType.LsstCam):` --> `if (defocalDist == 0) and (camType != CamType.LsstCam):`

**Review by teweitsai — APPROVED** [review]
Looks good and thanks for this update! Please help to double check the closed-loop simulation of ComCam and LsstFam can still run well. Thanks!

File: policy/phosimCmptSetting.yaml (hunk 1 above)
Comment (jbkalmbach) [reply policy/phosimCmptSetting.yaml:19]: This is where PhoSim works and saves the images within each `iterX/img` directory.

Comment (teweitsai) [reply policy/phosimCmptSetting.yaml:19]: Please help to add the description to the comment. Thanks!

**Comment by jbkalmbach** [comment]
Verified that this now runs with `comcam` and `lsstfam` mode as well.

**Pull request merged**